src/modules/claim/services/claim-verification.service.ts

```diff
@@ -0,0 +1,192 @@
+          `No credential found for role ${roleNamespace} for Did ${subjectDID}`,
+        ],
+      };
+    }
+    if (isEIP191Jwt(resolvedCredential)) {
+      return this.verifyRoleEIP191JWT(resolvedCredential as RoleEIP191JWT);
+    }
+
+    return {
+      isVerified: false,
+      errors: [`Verification not supported for resolved credential type`],
+    };
+  }
+
+  /**
+   * Verifies:
+   * - That off-chain claim was issued by authorized issuer
+   * - That claim is not expired
+   * - That off-chain claim proof is valid
+   *
+   * @param roleEIP191JWT off chain claim to verify
+   * @param subjectDID The credential subject
+   * @param roleNamespace The role to try to get a credential for. Should be a full role namespace (for example, "myrole.roles.myorg.auth.ewc")
+   * @return Boolean indicating if verified and array of error messages
+   */
+  private async verifyRoleEIP191JWT(
+    roleEIP191JWT: RoleEIP191JWT
+  ): Promise<{ isVerified: boolean; errors: string[] }> {
+    const { payload, eip191Jwt } = roleEIP191JWT;
+    const errors: string[] = [];
+    const issuerDID = roleEIP191JWT.payload?.iss;
+    if (!issuerDID) {
+      throw new Error('No issuer specified for credential');
+    }
+    const proofVerified = await this.verifyPublicClaim(
+      eip191Jwt,
+      payload?.iss as string
+    );
+    if (!proofVerified) {
+      errors.push(`Proof not verified for role`);
+    }
+    // Date.now() and JWT expiration time both identify the time elapsed since January 1, 1970 00:00:00 UTC
+    const isExpired = payload?.exp && payload?.exp * 1000 < Date.now();
+    if (isExpired) {
+      errors.push(`Credential for prerequisite role expired`);
+    }
+    const { verified: issuerVerified, error } =
+      await this.issuerVerificationService.verifyIssuer(
+        issuerDID,
+        payload?.claimData?.claimType
+      );
+    if (!issuerVerified && error) {
+      throw new Error(`Issuer verification failed: ${error}`);
+    }
+    return {
+      errors: errors,
+      isVerified: !!proofVerified && issuerVerified && !isExpired,
+    };
+  }
+
+  /**
+   * Verifies issued token of the public claim.
+   *
+   * @param {String} token JWT token of the public claim
+   * @param {String} iss DID of the issuer
+   * @return DID of the authenticated identity on successful verification or null otherwise
+   */
+  private async verifyPublicClaim(
+    token: string,
+    did: string
+  ): Promise<string | null> {
+    const didDoc = await this.didService.getById(did);
+    const verifier = new ProofVerifier(didDoc);
+    return verifier.verifyAssertionProof(token);
+  }
+
+  /**
+   * Verifies that a user's Did Document contains all roles required for enrolment (enrolment preconditions)
+   * @param userDID the Did of the user seeking to obtain the role. Must posses enrolment preconditions
+   * @param conditions enrolment preconditions needed to obtain the role (claimType)
+   */
+  public async verifyClaimPresentInDidDocument({
```

Comment:
Minor point, but I personally like having the public method of a class before the private methods. I proposed it as a code convention: https://energyweb.atlassian.net/wiki/spaces/MYEN/pages/3207626825/Code+Conventions+and+Guidelines

Reply:
@jrhender okay, I moved it to top above private mthds.

Reply:
Thanks. The intuition as that as someone reads a Class file, they are first interested in the public interface of the class and then they can continue reading if they want to know more implementation details